File: demo/android/Service/app/src/main/java/ai/picovoice/rhinodemoservice/RhinoService.java
/*
    Copyright 2018 Picovoice Inc.
    You may not use this file except in compliance with the license. A copy of the license is
    located in the "LICENSE" file accompanying this source.
    Unless required by applicable law or agreed to in writing, software distributed under the
    License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
    express or implied. See the License for the specific language governing permissions and
    limitations under the License.
*/

package ai.picovoice.rhinodemoservice;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat;

import java.io.File;
import java.util.Map;

import ai.picovoice.porcupine.PorcupineException;
import ai.picovoice.rhino.RhinoException;
import ai.picovoice.rhinomanager.AudioRecorder;

public class RhinoService extends Service {
    private static final String CHANNEL_ID = "PorcupineRhinoServiceChannel";

    private AudioRecorder audioRecorder;

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel notificationChannel = new NotificationChannel(
                    CHANNEL_ID,
                    "PorcupineRhinoServiceChannel",
                    NotificationManager.IMPORTANCE_HIGH);

            NotificationManager manager = getSystemService(NotificationManager.class);
            assert manager != null;
            manager.createNotificationChannel(notificationChannel);
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        createNotificationChannel();

        PendingIntent pendingIntent = PendingIntent.getActivity(
                this,
                0,
                new Intent(this, MainActivity.class),
                0);

        Notification notification = new NotificationCompat.Builder(this, CHANNEL_ID)
                .setContentTitle("Rhino")
                .setContentText("listening ...")
                .setSmallIcon(R.drawable.ic_launcher_background)
                .setContentIntent(pendingIntent)
                .build();

        startForeground(1234, notification);

        String porcupineModelFilePath = new File(this.getFilesDir(), "porcupine_params.pv").getAbsolutePath();
        String rhinoModelFilePath = new File(getFilesDir(), "rhino_params.pv").getAbsolutePath();

        String keywordFileName = intent.getStringExtra("keywordFileName");
        assert keywordFileName != null;
        String keywordFilePath = new File(getFilesDir(), keywordFileName).getAbsolutePath();


        String contextFileName = intent.getStringExtra("contextFileName");
        assert contextFileName != null;
        String contextFilePath = new File(this.getFilesDir(), contextFileName).getAbsolutePath();

        try {
            PorcupineRhinoAudioConsumer audioConsumer = new PorcupineRhinoAudioConsumer(
                    porcupineModelFilePath,
                    keywordFilePath,
                    rhinoModelFilePath,
                    contextFilePath,
                    (isWakeWordDetected, isUnderstood, rhinoIntent) -> {
                        CharSequence title = "PorcupineRhino";
                        PendingIntent contentIntent = PendingIntent.getActivity(
                                this,
                                0,
                                new Intent(this, MainActivity.class),
                                0);

                        String contextText;
                        if (isWakeWordDetected) {
                            contextText = "wake word detected. listening ...";
                        } else if (isUnderstood) {
                            contextText = "intent : " + rhinoIntent.getIntent() + " - ";
                            final Map<String, String> slots = rhinoIntent.getSlots();
                            for (String key : slots.keySet()) {
                                contextText += key + " : " + slots.get(key) + " - ";
                            }
                        } else {
                            contextText = "did not understand the command";
                        }

                        Notification n = new NotificationCompat.Builder(this, CHANNEL_ID)
                                .setContentTitle(title)
                                .setContentText(contextText)
                                .setSmallIcon(R.drawable.ic_launcher_background)
                                .setContentIntent(contentIntent)
                                .build();

                        NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
                        assert notificationManager != null;
                        notificationManager.notify(1234, n);
                    });

            audioRecorder = new AudioRecorder(audioConsumer);

            audioRecorder.start();
        } catch (RhinoException | PorcupineException e) {
            Log.e("PORCUPINE_SERVICE", e.toString());
        }

        return super.onStartCommand(intent, flags, startId);
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onDestroy() {
        try {
            if (audioRecorder != null) {
                audioRecorder.stop();
            }
        } catch (InterruptedException e) {
            //
        }

        super.onDestroy();
    }
}

File: binding/angular/projects/rhino-web-angular/src/lib/rhino.service.ts
/*
  Copyright 2022 Picovoice Inc.

  You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
  file accompanying this source.

  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
  specific language governing permissions and limitations under the License.
*/

import { Injectable, OnDestroy } from '@angular/core';
import { Subject } from 'rxjs';

import { WebVoiceProcessor } from '@picovoice/web-voice-processor';
import type {
  InferenceCallback,
  RhinoContext,
  RhinoOptions,
  RhinoInference,
  RhinoModel,
  RhinoWorker,
} from '@picovoice/rhino-web';

@Injectable({
  providedIn: 'root',
})
export class RhinoService implements OnDestroy {
  public contextInfo: string | null = null;
  public inference$: Subject<RhinoInference> = new Subject<RhinoInference>();
  public isLoaded: Subject<boolean> = new Subject<boolean>();
  public isListening: Subject<boolean> = new Subject<boolean>();
  public error$: Subject<Error | string | null> = new Subject<Error | string | null>();

  constructor() {}

  public async start(): Promise<boolean> {
    if (this.webVoiceProcessor !== null) {
      await this.webVoiceProcessor.start();
      this.listening$.next(true);
      return true;
    } else {
      return false;
    }
  }

  public async stop(): Promise<boolean> {
    if (this.webVoiceProcessor !== null) {
      await this.webVoiceProcessor.stop();
      this.listening$.next(false);
      return true;
    } else {
      return false;
    }
  }

  public async pushToTalk(): Promise<boolean> {
    if (!this.isTalking && this.rhinoWorker !== null) {
      this.isTalking = true;
      this.isTalking$.next(true);
      this.rhinoWorker.postMessage({ command: 'resume' });
      return true;
    }
    return false;
  }

  private InferenceCallback(inference: RhinoInference): void {
    this.inference$.next(inference)
  }

  public async init(
    accessKey: string,
    context: RhinoContext,
    model: RhinoModel,
    options: RhinoOptions = {}
  ): Promise<void> {
    if (this.isInit) {
      throw new Error('Rhino is already initialized');
    }
    const {
      accessKey,
      context,
      endpointDurationSec,
      requireEndpoint,
      start = true,
    } = rhinoServiceArgs;
    this.isInit = true;

    try {
      this.rhinoWorker = await rhinoWorkerFactory.create({
        accessKey,
        context,
        endpointDurationSec,
        requireEndpoint,
        start: false,
      });
      this.rhinoWorker.onmessage = (
        message: MessageEvent<RhinoWorkerResponse>
      ) => {
        switch (message.data.command) {
          case 'rhn-inference': {
            this.inference$.next(message.data.inference as RhinoInference);
            this.isTalking = false;
            this.isTalking$.next(false);
            break;
          }
          case 'rhn-info': {
            this.contextInfo = message.data.info;
            break;
          }
        }
      };
      this.rhinoWorker.postMessage({ command: 'info' });
    } catch (error) {
      this.isInit = false;
      this.isError$.next(true);
      this.error$.next(error as Error);
      throw error;
    }

    try {
      this.webVoiceProcessor = await WebVoiceProcessor.init({
        engines: [this.rhinoWorker],
        start,
      });
      this.listening$.next(start);
    } catch (error) {
      this.rhinoWorker.postMessage({ command: 'release' });
      this.rhinoWorker.terminate();
      this.rhinoWorker = null;
      this.isInit = false;
      this.isError$.next(true);
      this.error$.next(error as Error);
      throw error;
    }
  }

  async ngOnDestroy() {
    this.release();
  }
}
